File: scripts/generate_test_jwt.js
#!/usr/bin/env node
/**
 * 테스트용 JWT 토큰 생성 스크립트 (Node.js 내장 모듈만 사용)
 *
 * 사용법:
 *   node scripts/generate_test_jwt.js
 *   node scripts/generate_test_jwt.js --uuid <UUID>
 *   node scripts/generate_test_jwt.js --uuid <UUID> --secret <SECRET>
 *
 * 참고: 서버의 JwtUtil은 서명 검증 없이 payload만 디코딩합니다.
 *       (JwtUtil.java: Base64.getUrlDecoder().decode(parts[1]) 로 payload 추출)
 *       따라서 secret은 형식 유지를 위한 더미값으로 사용됩니다.
 */

const crypto = require("crypto");

function base64urlEncode(str) {
  return Buffer.from(str)
    .toString("base64")
    .replace(/=/g, "")
    .replace(/\+/g, "-")
    .replace(/\//g, "_");
}

function generateUUID() {
  return crypto.randomUUID();
}

function generateJwt(userUuid, secret = "test-secret", expiresInHours = 24) {
  const header = { alg: "HS256", typ: "JWT" };
  const now = Math.floor(Date.now() / 1000);
  const payload = {
    user_uuid: userUuid,
    iat: now,
    exp: now + expiresInHours * 3600,
  };

  const headerB64 = base64urlEncode(JSON.stringify(header));
  const payloadB64 = base64urlEncode(JSON.stringify(payload));
  const signingInput = `${headerB64}.${payloadB64}`;

  const signature = crypto
    .createHmac("sha256", secret)
    .update(signingInput)
    .digest("base64")
    .replace(/=/g, "")
    .replace(/\+/g, "-")
    .replace(/\//g, "_");

  return `${signingInput}.${signature}`;
}

// 인자 파싱
const args = process.argv.slice(2);
const getArg = (flag) => {
  const idx = args.indexOf(flag);
  return idx !== -1 ? args[idx + 1] : null;
};

const userUuid = getArg("--uuid") || generateUUID();
const secret = getArg("--secret") || "test-secret";
const expires = parseInt(getArg("--expires") || "24", 10);

const token = generateJwt(userUuid, secret, expires);

console.log("\n[user_uuid]");
console.log(`  ${userUuid}`);
console.log("\n[JWT Token]");
console.log(`  ${token}`);
console.log("\n[Authorization Header]");
console.log(`  Bearer ${token}`);
console.log("\n[curl 예시]");
console.log(`  curl -H "Authorization: Bearer ${token}" http://localhost:8084/api/...`);

File: scripts/generate_test_jwt.py
#!/usr/bin/env python3
"""
테스트용 JWT 토큰 생성 스크립트

사용법:
  python scripts/generate_test_jwt.py
  python scripts/generate_test_jwt.py --uuid <UUID>

참고: 서버의 JwtUtil은 서명 검증 없이 payload만 디코딩합니다.
      (JwtUtil.java: Base64.getUrlDecoder().decode(parts[1]) 로 payload 추출)
      따라서 secret은 형식 유지를 위한 더미값으로 사용됩니다.
"""

import base64
import json
import hmac
import hashlib
import argparse
import uuid as uuid_module
from datetime import datetime, timezone, timedelta


def base64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


def generate_jwt(user_uuid: str, secret: str = "test-secret", expires_in_hours: int = 24) -> str:
    header = {"alg": "HS256", "typ": "JWT"}
    now = datetime.now(timezone.utc)
    payload = {
        "user_uuid": user_uuid,
        "iat": int(now.timestamp()),
        "exp": int((now + timedelta(hours=expires_in_hours)).timestamp()),
    }

    header_b64 = base64url_encode(json.dumps(header, separators=(",", ":")).encode())
    payload_b64 = base64url_encode(json.dumps(payload, separators=(",", ":")).encode())

    signing_input = f"{header_b64}.{payload_b64}"
    signature = hmac.new(secret.encode(), signing_input.encode(), hashlib.sha256).digest()
    signature_b64 = base64url_encode(signature)

    return f"{signing_input}.{signature_b64}"


def main():
    parser = argparse.ArgumentParser(description="테스트용 JWT 토큰 생성")
    parser.add_argument("--uuid", default=str(uuid_module.uuid4()), help="user_uuid 값 (미입력시 랜덤 생성)")
    parser.add_argument("--secret", default="test-secret", help="JWT 서명용 시크릿 (기본값: test-secret)")
    parser.add_argument("--expires", type=int, default=24, help="만료 시간(시간 단위, 기본값: 24)")
    args = parser.parse_args()

    token = generate_jwt(args.uuid, args.secret, args.expires)

    print(f"\n[user_uuid]")
    print(f"  {args.uuid}")
    print(f"\n[JWT Token]")
    print(f"  {token}")
    print(f"\n[Authorization Header]")
    print(f"  Bearer {token}")
    print(f"\n[curl 예시]")
    print(f'  curl -H "Authorization: Bearer {token}" http://localhost:8084/api/...')


if __name__ == "__main__":
    main()
